internal/crypto/src/time_stamp/verify.rs

```diff
@@ -0,0 +1,368 @@
+// Copyright 2022 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use std::ops::Deref;
+
+use async_generic::async_generic;
+use bcder::{decode::SliceSource, encode::Values, ConstOid, OctetString};
+use chrono::{offset::LocalResult, DateTime, TimeZone, Utc};
+use x509_certificate::{
+    asn1time::{GeneralizedTime, GeneralizedTimeAllowedTimezone, Time},
+    DigestAlgorithm,
+};
+
+use crate::{
+    asn1::{
+        rfc3161::TstInfo,
+        rfc5652::{
+            CertificateChoices::Certificate, SignerIdentifier, OID_MESSAGE_DIGEST, OID_SIGNING_TIME,
+        },
+    },
+    raw_signature::validator_for_sig_and_hash_algs,
+    time_stamp::{TimeStampError, TimeStampResponse},
+};
+
+/// Decode the TimeStampToken info and verify it against the supplied data.
+///
+/// TEMPORARILY PUBLIC while refactoring
+#[async_generic]
+pub fn verify_time_stamp(ts: &[u8], data: &[u8]) -> Result<TstInfo, TimeStampError> {
+    let ts_resp = TimeStampResponse::parse(ts)?;
+
+    // Did the time stamp expire between issuance and verification?
+    let Some(sd) = &ts_resp.signed_data()? else {
+        return Err(TimeStampError::DecodeError(
+            "unable to find signed data".to_string(),
+        ));
+    };
+
+    let certs = sd.certificates.clone().ok_or(TimeStampError::DecodeError(
+        "time stamp contains no certificates".to_string(),
+    ))?;
+
+    let mut last_err = TimeStampError::InvalidData;
+
+    // Look for any valid signer.
+    for signer_info in sd.signer_infos.iter() {
+        // Find signer's cert.
+        let cert = match certs.iter().find_map(|cc| {
+            let c = match cc {
+                Certificate(c) => c,
+                _ => return None,
+            };
+
+            match &signer_info.sid {
```

Comment:
Move these to separate functions that clearly define their purpose

Reply:
No, why.  This is not separate functionality.  It is doing a specific process.

Reply:
Won't implement in this PR. Tracking as part of CAI-7310.